File: 1325-delete-leaves-with-a-given-value/1325-delete-leaves-with-a-given-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* removeLeafNodes_01(TreeNode* root, int target) {
        if(root == nullptr) return nullptr;
        
        root->left = removeLeafNodes_01(root->left,target);
        
        root->right = removeLeafNodes_01(root->right,target);
        
        if(root->left == nullptr && root->right == nullptr && root->val == target) return nullptr;
        else return root;
    }
    
    TreeNode* removeLeafNodes(TreeNode* root, int target) {
        if(root == nullptr || (root->val == target && root->left == nullptr && root->right == nullptr)) return nullptr;
      
        return removeLeafNodes_01(root,target);
    }
};

File: 1325-delete-leaves-with-a-given-value/1325-delete-leaves-with-a-given-value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode removeLeafNodes_01(TreeNode node, int target) {
        //Base Condition
        if(node == null) return null;
        // Recursive Calls
        node.left = removeLeafNodes_01(node.left,target);
        node.right = removeLeafNodes_01(node.right,target);
        
        //PostOrder checking of leafs
        if(node.left == null && node.right == null && node.val == target) return null;
        
        return node;
    }
    
    public TreeNode removeLeafNodes(TreeNode root, int target) {
        if(root == null) return null;
        return removeLeafNodes_01(root,target);
    }
}